### DISCUSSION TITLE
Training skills that lack a lot of recipes via training the secondary skills

### DISCUSSION THREAD

**Discussion by ghost** [cataclysmbn/Cataclysm-BN#3825] [General]
Right now, secondary skills don't gain xp during a craft and are mainly there for autolearn and increased failure chance if you don't meet them. Allowing secondary skill requirements to train a skill would make some skills such as mechanics easier to train. First referenced in https://github.com/cataclysmbnteam/Cataclysm-BN/pull/3800

**Comment by scarf005** [comment]
i agree, we could reference [DCSS's cross training](http://crawl.chaosforge.org/Cross-train)

**Comment by chaosvolt** [comment]
If anyone can point to what function in the code is responsible for calling the practice function while crafting that would be a big help, I've attempted to find it in the past (for the sake of trying to sus out the rate at which crafting generates EXP per hour so I have a benchmark for what other activities should be doing to compete with craft-grinding) and failed to find it.

**Reply by scarf005** [reply]
## Fields

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/04b463a6a7d78c28e5a228deeb258166f3e43895/src/recipe.h#L111-L112

`skill_used`: primary skill
`required_skills`: map of secondary skills

## Crafting success chance

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/04b463a6a7d78c28e5a228deeb258166f3e43895/src/crafting.cpp#L816-L823

## Skill gain from crafting function

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/04b463a6a7d78c28e5a228deeb258166f3e43895/src/crafting.cpp#L770

### Practicing Primary Skill

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/04b463a6a7d78c28e5a228deeb258166f3e43895/src/crafting.cpp#L790

**Reply by chaosvolt** [reply]
Ayy, good shit :D

**Reply by chaosvolt** [reply]
So I assume it would be added in `player::craft_skill_gain` and the resulting format would look roughly like this:
```C++
     for( const auto &pr : making.required_skills ) { 
         practice( pr, base_practice * 0.5, skill_cap, true ); 
     } 
```

**Reply by scarf005** [reply]
i think so, but the inner logic should be tidied up first into multiple functions.

**Comment by scarf005** [comment]
i this discussion could be closed as concrete solutions are made and accepted.
- #3921
- #3932
